Download images into pro local backups

Quiescent opened the pull request:
- Re-instate the firebase backup mechanism.
 - Ensure that local backups of pro files have the correct name.
 - Always backup a self-contained version of the plottr file (i.e. download all `storage://` protocol items back into the file before saving it.)

Quiescent committed:
Add call to save backups on frb & download images into local backups

Quiescent committed:
Fix the names of backup files

Quiescent committed:
Export the start of session backup to a self-contained file

Quiescent committed:
Fix failing tests

jeanawhou reviewed (approved):
LGTM :)

cameronsutter reviewed (commented):
I could be reading this wrong, but my biggest concern here is that we're downloading the images for every backup, right?

Can we cache them locally somehow? Those aren't going to change often … except when they do of course

Quiescent commented:
> I could be reading this wrong, but my biggest concern here is that we're downloading the images for every backup, right?
> 
> Can we cache them locally somehow? Those aren't going to change often … except when they do of course

You're right.  This is why we didn't have this feature in the first place. :thinking: 
I think it's safe to have a per-session in-memory cache.
I'll start work on that now.
We can always change it to a disk-level cache if we have to, but in-memory is easier, to begin with.

Quiescent committed:
Add an in-memory cached image download mechanism

Quiescent committed:
Use the cached version of storage image downloader

Quiescent commented:
> > I could be reading this wrong, but my biggest concern here is that we're downloading the images for every backup, right?
> > Can we cache them locally somehow? Those aren't going to change often … except when they do of course
> 
> You're right. This is why we didn't have this feature in the first place. thinking I think it's safe to have a per-session in-memory cache. I'll start work on that now. We can always change it to a disk-level cache if we have to, but in-memory is easier, to begin with.

@cameronsutter , I've implemented a cached version.  WDYT?

cameronsutter reviewed (approved):
LGTM

Quiescent committed:
Fix the `backupFile` tests

They needed an extra (dummy) argument for downlnoading images back
into the file.
